core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -157,7 +157,9 @@ private Type findFieldType(int fieldId) {
 
   private void addColumn(int parentId, Types.NestedField field) {
     String parentName = partnerSchema.findColumnName(parentId);
-    api.addColumn(parentName, field.name(), field.type(), field.doc());
+    String fullName = (parentName != null ? parentName + "." : "") + field.name();
+    api.addColumn(parentName, field.name(), field.type(), field.doc(), field.initialDefault())
+        .updateColumnDefault(fullName, field.writeDefault());
```

Comment:
To preserve both the initial and write defaults, this creates the field with the initial default (which sets both) and then updates the field, which sets only the write default because the initial default cannot change once it is set.